File: CS50x-Week6-Solution-With-Deep-Explanation-main/cash (feeling less comfortable)/cash.py
# Import the get_float function from the cs50 library
from cs50 import get_float

# Continuously prompt the user for input until a valid non-negative value is provided
while True:
    # Use get_float to prompt the user for the amount of change owed
    # This function displays the prompt message and waits for the user to enter a floating-point number
    change = get_float("Change owed: ")
    
    # Check if the provided amount of change is non-negative
    if change >= 0:
        # Exit the loop if a valid (non-negative) input is received
        break
    else:
        # If the input is negative, print an error message and prompt the user again
        print("Invalid input. Please enter a non-negative value.")

# Convert the amount of change from dollars to cents
# This is done to handle the change in integer values (cents) for easier coin calculations
# Multiply by 100 and round to the nearest whole number to avoid floating-point precision issues
cents = round(change * 100)

# Define the values of the different coin denominations in cents
quarters = 25  # Value of one quarter in cents
dimes = 10     # Value of one dime in cents
nickels = 5    # Value of one nickel in cents
pennies = 1    # Value of one penny in cents

# Initialize a counter to keep track of the total number of coins used
num_coins = 0

# Calculate the number of quarters needed
# Use integer division to determine how many quarters fit into the total cents
num_coins += cents // quarters
# Update the remaining cents after accounting for quarters
cents %= quarters

# Calculate the number of dimes needed
# Use integer division to determine how many dimes fit into the remaining cents
num_coins += cents // dimes
# Update the remaining cents after accounting for dimes
cents %= dimes

# Calculate the number of nickels needed
# Use integer division to determine how many nickels fit into the remaining cents
num_coins += cents // nickels
# Update the remaining cents after accounting for nickels
cents %= nickels

# Calculate the number of pennies needed
# Use integer division to determine how many pennies fit into the remaining cents
num_coins += cents // pennies

# Output the total number of coins needed to make the change
print(num_coins)

File: CS50x-Week1-Solution-With-Deep-Explanation-main/cash (feeling less comfortable)/cash.c
#include <cs50.h> // This library is used to leverage the get_int function, which helps in safely getting integer input from the user.
#include <stdio.h> // This library is used for input and output functions like printf.

int main(void) // function is the entry point of the program.
{
    // Variable to store the amount of money (in cents) that the user owes as change.
    int money;
    
    // Prompt the user to enter a non-negative amount of change owed.
    // This loop continues until a non-negative integer is entered.
    do
    {
        // Get the change owed from the user.
        money = get_int("Change owed: ");
    } while (money < 0); // Repeat the loop if the entered value is negative.

    // Calculate the number of quarters (25 cents each).
    int quarters = money / 25;
    // Update the remaining amount of money after using quarters.
    money %= 25;

    // Calculate the number of dimes (10 cents each).
    int dimes = money / 10;
    // Update the remaining amount of money after using dimes.
    money %= 10;

    // Calculate the number of nickels (5 cents each).
    int nickels = money / 5;
    // Update the remaining amount of money after using nickels.
    money %= 5;

    // The remaining money is entirely in pennies (1 cent each).
    int pennies = money;
    
    // Calculate the total number of coins used.
    int total_coins = quarters + dimes + nickels + pennies;
    
    // Print the total number of coins required to make the change.
    printf("%d\n", total_coins);
}
